lucene/core/src/java/org/apache/lucene/document/IndexSortDocValuesRangeQuery.java

```diff
@@ -0,0 +1,279 @@
+              "on the same field as the query field.");
+        }
+
+        SortedNumericDocValues sortedDocValues = reader.getSortedNumericDocValues(field);
+        NumericDocValues docValues = DocValues.unwrapSingleton(sortedDocValues);
+        if (docValues == null) {
+          throw new IllegalArgumentException("To use IndexSortDocValuesRangeQuery, the query field must have " +
+              "SortedNumericDocValues and have at most one value per document.");
+        }
+
+        DocIdSetIterator disi = getDocIdSetIterator(lowerValue, upperValue,
+            sortField, context, docValues);
+        return new ConstantScoreScorer(this, score(), scoreMode, disi);
+      }
+
+      @Override
+      public boolean isCacheable(LeafReaderContext ctx) {
+        return DocValues.isCacheable(ctx, field);
+      }
+    };
+  }
+
+  /**
+   * Computes the document IDs that lie within the range [lowerValue, upperValue] by
+   * performing binary search on the field's doc values.
+   *
+   * Because doc values only allow forward iteration, we need to reload the field comparator
+   * every time the binary search accesses an earlier element.
+   */
+  private static DocIdSetIterator getDocIdSetIterator(long lowerValue,
+                                                      long upperValue,
+                                                      SortField sortField,
+                                                      LeafReaderContext context,
+                                                      DocIdSetIterator delegate) throws IOException {
+    int maxDoc = context.reader().maxDoc();
+    if (maxDoc <= 0) {
+      return DocIdSetIterator.empty();
+    }
+
+    if (sortField.getReverse()) {
+      long temp = lowerValue;
+      lowerValue = upperValue;
+      upperValue = temp;
+    }
+
+    // Perform a binary search to find the first document with value >= lowerValue.
+    ValueComparator comparator = loadComparator(sortField, lowerValue, context);
+    int low = 0;
+    int high = maxDoc - 1;
+
+    while (low <= high) {
+      int mid = (low + high) >>> 1;
+      if (comparator.compare(mid) <= 0) {
+        high = mid - 1;
+        comparator = loadComparator(sortField, lowerValue, context);
+      } else {
+        low = mid + 1;
+      }
+    }
+    int firstDocIdInclusive = high + 1;
+
+    // Perform a binary search to find the first document with value > upperValue.
+    // Since we know that upperValue >= lowerValue, we can initialize the lower bound
+    // of the binary search to the result of the previous search.
+    comparator = loadComparator(sortField, upperValue, context);
+    low = firstDocIdInclusive;
+    high = maxDoc - 1;
+
+    while (low <= high) {
+      int mid = (low + high) >>> 1;
+      if (comparator.compare(mid) < 0) {
+        high = mid - 1;
+        comparator = loadComparator(sortField, upperValue, context);
```

Comment:
As mentioned above, can we ensure the cost of this iteration does not override the benefits?